File: LeetCode/src/backtracking/WordSearch.java
package backtracking;

public class WordSearch {
    public boolean exist(char[][] board, String word) {

        int rows = board.length;
        int cols = board[0].length;

        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (backtracking(board, word, i, j, 0)) {
                    return true;
                }
            }
        }

        return false;
    }

    private boolean backtracking(char[][] board, String word, int row, int col, int searchIndex) {
        if (searchIndex == word.length()) {
            return true;
        }

        int rows = board.length;
        int cols = board[0].length;

        if (row < 0 || row >= rows || col < 0 || col >= cols || board[row][col] != word.charAt(searchIndex)) {
            return false;
        }

        board[row][col] = '#';

        boolean result = backtracking(board, word, row - 1, col, searchIndex + 1) ||
                backtracking(board, word, row + 1, col, searchIndex + 1) ||
                backtracking(board, word, row, col - 1, searchIndex + 1) ||
                backtracking(board, word, row, col + 1, searchIndex + 1);

        board[row][col] = word.charAt(searchIndex);

        return result;
    }

    public static void main(String[] args) {
        WordSearch solution = new WordSearch();
        char[][] board = {
                {'A', 'B', 'C', 'E'},
                {'S', 'F', 'C', 'S'},
                {'A', 'D', 'E', 'E'}
        };
        String word1 = "ABCCED";
        String word2 = "SEE";
        String word3 = "ABCB";

        System.out.println(solution.exist(board, word1)); // Output: true
        System.out.println(solution.exist(board, word2)); // Output: true
        System.out.println(solution.exist(board, word3)); // Output: false
    }
}

File: LeetCode_Python/backtracking/WordSearch.py
class WordSearch:
    def exist(self, board: list[list[str]], word: str) -> bool:
        m, n = len(board), len(board[0])

        def backtracking(index: int, row: int, col: int) -> bool:
            if index == len(word):
                return True

            if (
                row < 0
                or row >= m
                or col < 0
                or col >= n
                or board[row][col] != word[index]
            ):
                return False

            board[row][col] = "#"

            found = (
                backtracking(index + 1, row - 1, col)
                or backtracking(index + 1, row + 1, col)
                or backtracking(index + 1, row, col - 1)
                or backtracking(index + 1, row, col + 1)
            )

            board[row][col] = word[index]

            return found

        for i in range(m):
            for j in range(n):
                if backtracking(0, i, j):
                    return True

        return False


if __name__ == "__main__":
    solution = WordSearch()

    board = [["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]]
    word = "ABCCED"
    print("Expected Output : True")
    print("Actual Output : ", solution.exist(board, word))

    board = [["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]]
    word = "SEE"
    print("Expected Output : True")
    print("Actual Output : ", solution.exist(board, word))

    board = [["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]]
    word = "ABCB"
    print("Expected Output : False")
    print("Actual Output : ", solution.exist(board, word))
